app/controllers/library_hours_controller.rb

```diff
@@ -1,24 +1,108 @@
 # frozen_string_literal: true
 
 class LibraryHoursController < ApplicationController
+  before_action :get_locations, only: [:index, :show]
+  before_action :set_location, only: [:show]
+  before_action :set_dates, only: [:index, :show]
+
   def index
-    @locations = LibraryHours.all.pluck(:location_id).uniq
+    @buildings = [
+      {
+        slug: "paley",
+        spaces: [
+                  "paley",
+                  "media",
+                  "doc_del",
+                  "ref_desk",
+                  "v_ref",
+                  "thinktank",
+                  "scrc",
+                  "dsc",
+                  "guest_computers"
+                ]
+      },
+      {
+        slug: "ginsburg",
+        spaces: ["ginsburg", "innovation"]
+      },
+      {
+        slug: "podiatry",
+        spaces: ["podiatry"]
+      },
+      {
+        slug: "ambler",
+        spaces: ["ambler"]
+      },
+      {
+        slug: "blockson",
+        spaces: ["blockson"]
+      }
+    ]
+    @buildings.each do |building|
+      building.values.second.map! do |space|
+        space = [building.values.first, LibraryHours.where(location_id: space, date: @sunday..@saturday)]
+      end
+    end
   end
-  def show
-    @locations = LibraryHours.distinct.pluck(:location_id)
+
+  def set_dates
     @today = Date.today
-    sunday = @today.beginning_of_week - 1
-    saturday = @today.end_of_week
+    @cyear = @today.year
+    @cweek = @today.cweek
+
+    @week =  params[:week].nil? ? @cweek : params[:week].to_i
+    @year = params[:year].nil? ? @cyear : params[:year].to_i
+
+    @first_week = LibraryHours.where(location_id: "paley").first.date.to_date.cweek
```

Comment:
We're doing four database queries here on every page, even if we don't need to know about next week. It feels like we should only calculate that on demand, not every time